lib/Builtins.cpp

```diff
@@ -273,113 +273,117 @@ Builtins::Lookup(const std::string &mangled_name) {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-////  Legacy interface
-////////////////////////////////////////////////////////////////////////////////
-bool Builtins::IsImageBuiltin(StringRef name) {
-  auto func_type = Lookup(name).getType();
-  return func_type > kType_Image_Start && func_type < kType_Image_End;
-}
-
-bool Builtins::IsSampledImageRead(StringRef name) {
-  return IsFloatSampledImageRead(name) || IsUintSampledImageRead(name) ||
-         IsIntSampledImageRead(name);
-}
-
-bool Builtins::IsFloatSampledImageRead(StringRef _name) {
-  const auto &fi = Lookup(_name);
-  if (fi.getType() == kReadImagef) {
-    const auto &pi = fi.getParameter(1);
-    return pi.isSampler();
+// Generate a mangled name loosely based on Itanium mangling
+std::string Builtins::GetMangledFunctionName(const char *name, Type *type) {
+  assert(name);
+  std::string mangled_name =
+      std::string("_Z") + std::to_string(strlen(name)) + name;
+  if (auto *func_type = dyn_cast<FunctionType>(type)) {
+    Type *last_arg_type = nullptr;
+    for (auto *arg_type : func_type->params()) {
+      if (arg_type == last_arg_type) {
+        mangled_name += "S_";
+      } else {
+        mangled_name += GetMangledTypeName(arg_type);
+        last_arg_type = arg_type;
+      }
+    }
+  } else {
+    mangled_name += GetMangledTypeName(type);
   }
-  return false;
+  return mangled_name;
 }
 
-bool Builtins::IsUintSampledImageRead(StringRef _name) {
-  const auto &fi = Lookup(_name);
-  if (fi.getType() == kReadImageui) {
-    const auto &pi = fi.getParameter(1);
-    return pi.isSampler();
-  }
-  return false;
+// The mangling follows the Itanium convention.
+std::string Builtins::GetMangledFunctionName(const char *name) {
+  assert(name);
+  return std::string("_Z") + std::to_string(strlen(name)) + name;
 }
 
-bool Builtins::IsIntSampledImageRead(StringRef _name) {
-  const auto &fi = Lookup(_name);
-  if (fi.getType() == kReadImagei) {
-    const auto &pi = fi.getParameter(1);
-    return pi.isSampler();
-  }
-  return false;
-}
+// The mangling loosely follows the Itanium convention.
+// Its purpose is solely to ensure uniqueness of names, it is not
+// meant to convey type information.
+std::string Builtins::GetMangledTypeName(Type *Ty) {
+  std::string mangled_type_str;
+
+  switch (Ty->getTypeID()) {
+  case Type::VoidTyID:
+    return "v";
+  case Type::HalfTyID:
+  case Type::FloatTyID:
+  case Type::DoubleTyID:
+    switch (Ty->getScalarSizeInBits()) {
```

Comment:
This seems like it could be simplified to:
```
case Type::HalfTyID:
  return "Dh";
case Type::FloatTyID:
  return "f";
case Type::DoubleTyID:
  return "d";
```

Reply:
Yes, true, since this is OpenCL kernel language I guess we can rely on the type.